Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewSnapshot.mm

```diff
@@ -717,4 +752,60 @@ - (void)takeSnapshotWithWebView:(WKWebView *)webView configuration:(WKSnapshotCo
 }
 #endif
 
+TEST(WKWebView, SnapshotNodeByJSHandle)
+{
+    RetainPtr webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 400, 400)]);
+    [webView synchronouslyLoadTestPageNamed:@"node-snapshotting"];
+
+    RetainPtr<Util::PlatformWindow> window;
+
+#if PLATFORM(MAC)
+    window = adoptNS([[NSWindow alloc] initWithContentRect:[webView frame] styleMask:NSBorderlessWindowMask backing:NSBackingStoreBuffered defer:NO]);
+    [[window contentView] addSubview:webView.get()];
+#elif PLATFORM(IOS_FAMILY)
+    window = adoptNS([[UIWindow alloc] initWithFrame:[webView frame]]);
+    [window addSubview:webView.get()];
+#endif
+
+    RetainPtr worldConfiguration = adoptNS([_WKContentWorldConfiguration new]);
+    [worldConfiguration setAllowJSHandleCreation:YES];
+
+    RetainPtr world = [WKContentWorld _worldWithConfiguration:worldConfiguration.get()];
+    auto querySelector = [&](ASCIILiteral selector) {
+        RetainPtr script = [NSString stringWithFormat:@"window.webkit.jsHandle(document.querySelector('%s'))", selector.characters()];
+        return dynamic_objc_cast<_WKJSHandle>([webView objectByEvaluatingJavaScript:script.get() inFrame:nil inContentWorld:world.get()]);
+    };
+
+    RetainPtr redBoxElement = querySelector(".red-box"_s);
+    RetainPtr imageElement = querySelector("img"_s);
+    RetainPtr headingElement = querySelector("h2"_s);
+
+    {
+        auto [image, error] = [webView takeSnapshotOfNode:redBoxElement.get()];
+        EXPECT_NULL(error);
+
+        CGImagePixelReader reader { Util::convertToCGImage(image.get()).get() };
+        EXPECT_WK_STREQ("rgb(255, 0, 0)", reader.cssColorAt(2, 2));
+        EXPECT_WK_STREQ("rgb(255, 0, 0)", reader.cssColorAt(reader.width() / 2, reader.height() / 2));
+    }
+
+    {
+        auto [image, error] = [webView takeSnapshotOfNode:imageElement.get()];
+        EXPECT_NULL(error);
+
+        CGImagePixelReader reader { Util::convertToCGImage(image.get()).get() };
+        EXPECT_TRUE(reader.isTransparentBlack(2, 2));
+        EXPECT_FALSE(reader.isTransparentBlack(reader.width() / 2, reader.height() / 2));
+    }
+
+    {
+        auto [image, error] = [webView takeSnapshotOfNode:headingElement.get()];
+        EXPECT_NULL(error);
+
+        CGImagePixelReader reader { Util::convertToCGImage(image.get()).get() };
+        EXPECT_WK_STREQ("rgb(0, 0, 0)", reader.cssColorAt(2, 2));
+        EXPECT_FALSE(reader.isTransparentBlack(reader.width() / 2, reader.height() / 2));
+    }
```

Comment:
Maybe coverage for an error case? Pass in a null handle?

Reply:
Yes, will add!